ir/instr.cpp

```diff
@@ -1865,4 +1865,121 @@ unique_ptr<Instr> Memcpy::dup(const string &suffix) const {
   return make_unique<Memcpy>(*dst, *src, *bytes, align_dst, align_src, move);
 }
 
+vector<Value*> ExtractElement::operands() const {
+  return { a, idx };
+}
+
+void ExtractElement::rauw(const Value &what, Value &with) {
+  RAUW(a);
+  RAUW(idx);
+}
+
+void ExtractElement::print(ostream &os) const {
+  os << getName() << " = extractelement " << *a << ", " << *idx;
+}
+
+StateValue ExtractElement::toSMT(State &s) const {
+  auto &[iv, ip] = s[*idx];
+  auto aty = a->getType().getAsAggregateType();
+  expr inbounds = iv.ult(aty->numElementsConst());
+  auto [rv, rp] = aty->extract(s[*a], iv);
+  return { move(rv), ip && inbounds && rp };
+}
+
+expr ExtractElement::getTypeConstraints(const Function &f) const {
+  return Value::getTypeConstraints() &&
+         a->getType().enforceVectorType([&](auto &ty) { return ty == getType(); }) &&
+         idx->getType().enforceIntType();
+}
+
+unique_ptr<Instr> ExtractElement::dup(const string &suffix) const {
+  return make_unique<ExtractElement>(getType(), getName() + suffix, *a, *idx);
+}
+
+
+vector<Value*> InsertElement::operands() const {
+  return { a, e, idx };
+}
+
+void InsertElement::rauw(const Value &what, Value &with) {
+  RAUW(a);
+  RAUW(e);
+  RAUW(idx);
+}
+
+void InsertElement::print(ostream &os) const {
+  os << getName() << " = insertelement" << *a << ", " << *e << ", " << *idx;
+}
+
+StateValue InsertElement::toSMT(State &s) const {
+  auto &[iv, ip] = s[*idx];
+  auto aty = a->getType().getAsAggregateType();
+  expr inbounds = iv.ult(aty->numElementsConst());
+  auto [rv, rp] = aty->update(s[*a], s[*e], iv);
+  return { move(rv), expr::mkIf(ip && inbounds,
+                                move(rp), expr::mkInt(-1, aty->bits())) };
```

Comment:
this breaks the abstraction.
I guess the update function should do this, since anyway the case of what happens on idx overflow needs to be considered.

Reply:
The the extract() from AggregateTypes does not consider idx overflow cases, either. I was just following that.

Reply:
If you prefer handle the o.o.b. cases inside update, I can fix that :)

Reply:
Right, for extract things were a bit easier because you get a single element, while for update you can't do the same.
Should we move the OOB check in both cases to their respective functions?  What do you think?
I really don't like exposing the BV representation since we certainly will want to play with different variations and we want everything in one place to avoid bugs.

Reply:
Ah, never mind. Better suggestion: you can use Type::getDummyValue(false).non_poison to get your all-poison value.

Reply:
I am more concern about what will be the layer for the pluggable UB support work upon? If we're planning to make the pluggable UB works at instr level, then handling OOB outside update() may be more reasonable.